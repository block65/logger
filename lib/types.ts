import { AsyncLocalStorage } from 'async_hooks';
import type pino from 'pino';
import type { MixinFn, MixinFnWithData } from './mixins.js';

export type Falsy = false | undefined | null;

export enum LogLevelNumbers {
  Fatal = 60,
  Error = 50,
  Warn = 40,
  Info = 30,
  Debug = 20,
  Trace = 10,
  Silent = 0,
}

export interface LogDescriptor {
  level: LogLevelNumbers;
  time: number;

  msg?: string;
  name?: string | undefined;

  // maye not be available depending on platform/cli
  pid?: number;
  hostname?: string;

  [key: string]: unknown;
}

type RemoveIndex<T> = {
  [K in keyof T as {} extends Record<K, 1> ? never : K]: T[K];
};

// pino tacks on a nasty `Record<string, any>` to its logger type
// we try to remove it here
export type BaseLogger = RemoveIndex<pino.Logger>;

export interface Logger extends BaseLogger {
  als: AsyncLocalStorage<AlsContext>;
}

export interface AlsContext {
  id: string;
  context?: Record<string, unknown>;
}

export interface AlsContextOutput {
  _contextId: AlsContext['id'];
  _context?: AlsContext['context'];
}

export type ComputePlatform = 'gcp-cloudrun' | 'aws-lambda' | 'aws';

export interface CreateLoggerOptions
  extends Omit<pino.LoggerOptions, 'mixin' | 'prettyPrint' | 'prettifier'> {
  pretty?: boolean;
  traceCaller?: boolean;
  platform?: ComputePlatform;
  mixins?: (MixinFnWithData | MixinFn | Falsy)[];
  transports?: pino.TransportMultiOptions;
}

export type CreateLoggerOptionsWithoutTransports = Omit<
  CreateLoggerOptions,
  'pretty' | 'transports'
>;

export interface CreateCliLoggerOptions
  extends Omit<CreateLoggerOptionsWithoutTransports, 'platform'> {
  color?: boolean;
}
