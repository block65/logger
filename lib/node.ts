import type { RequestHandler } from 'express';
import { serializeError } from 'serialize-error';
import * as pino from 'pino';
import * as pinoHttp from 'pino-http';
import { createNamespace, Namespace } from 'cls-hooked';
import { hostname } from 'os';
import { randomBytes } from 'crypto';
import { IncomingMessage, ServerResponse } from 'http';
import { sep } from 'path';

type Falsy = false | undefined | null;
type MixinFnWithData = (
  data: ReturnType<pino.MixinFn>,
) => ReturnType<pino.MixinFn>;

export type Logger = {
  cls: Namespace;
} & pino.Logger;

interface ClsContext {
  _contextId: string;
  _context?: Record<string, unknown>;
}

type NamespaceContext = {
  id?: unknown;
  _ns_name?: unknown;
} & Partial<ClsContext>;

type ComputePlatform = 'gcp-cloudrun' | 'aws-lambda' | 'aws';

export interface CreateLoggerOptions extends Omit<pino.LoggerOptions, 'mixin'> {
  traceCaller?: boolean;
  platform?: ComputePlatform;
  mixins?: (MixinFnWithData | pino.MixinFn | Falsy)[];
}

const suggestedHostname = hostname();

const defaultLoggerOptions: pino.LoggerOptions = {
  level: 'info',
  serializers: {
    err: serializeError,
  },
  base: {
    pid: process.pid,
    // in the case we get localhost (gcp cloud run), use random bytes instead
    // this helps us debug if we need to isolate calls running in a specific
    // container or instance (one which stays alive between invocations)
    hostname:
      suggestedHostname !== 'localhost'
        ? suggestedHostname
        : `zz${randomBytes(5).toString('hex')}`, // zz indicates synthetic hostname
  },
};

const gcpLevelMap: Record<string, string> = {
  trace: 'DEFAULT',
  debug: 'DEBUG',
  info: 'INFO',
  warn: 'WARNING',
  error: 'ERROR',
  fatal: 'CRITICAL',
};

function getPlatformLoggerOptions(
  platform?: ComputePlatform,
): pino.LoggerOptions {
  switch (platform) {
    case 'gcp-cloudrun':
      return {
        level: 'info',
        // customLevels: {
        //   DEBUG: 100,
        //   INFO: 200,
        //   NOTICE: 300,
        //   WARNING: 400,
        //   ERROR: 500,
        //   CRITICAL: 600,
        //   ALERT: 700,
        //   EMERGENCY: 800,
        // },
        messageKey: 'message',
        formatters: {
          level(levelLabel /* , levelNumber */) {
            return { severity: gcpLevelMap[levelLabel] };
          },
        },
        timestamp: pino.stdTimeFunctions.isoTime,
      };
    case 'aws':
    case 'aws-lambda':
    default:
      return {};
  }
}

let counter = 0;

const isDevelopment = process.env.NODE_ENV === 'development';

function callerMixin(): { caller: string | undefined } {
  return {
    caller: Error()
      .stack?.split('\n')
      .slice(2)
      .filter(
        (s) =>
          // !s.includes(`api${sep}src${sep}app`) &&
          // !s.includes(
          //   `node_modules${sep}express${sep}lib${sep}router`,
          // ) &&
          !s.includes(`node_modules${sep}pino`) &&
          !s.includes(`block65${sep}logger`),
      )?.[0]
      ?.substr(7),
  };
}

function composeMixins(
  mixins: (MixinFnWithData | pino.MixinFn | Falsy)[],
): pino.MixinFn {
  return () =>
    mixins.reduce((accum, mixin) => {
      if (!mixin) {
        return accum;
      }
      return {
        ...accum,
        ...mixin(accum),
      };
    }, {});
}

function createContextMixin(cls: Namespace): pino.MixinFn {
  return (): Partial<ClsContext> => {
    const { active }: { active: NamespaceContext | null } = cls || {};

    // eslint-disable-next-line @typescript-eslint/naming-convention
    const { id, _ns_name, ...clsContext } = active || {};
    return clsContext;
  };
}

export function createLogger(
  opts: CreateLoggerOptions = {},
  stream?: pino.DestinationStream,
  // destination?: pino.DestinationStream,
): Logger {
  const cls = createNamespace(`logger/${counter}`);
  counter += 1;

  const {
    platform,
    traceCaller = isDevelopment,
    mixins = [],
    ...userPinoOpts
  } = opts;

  const platformLoggerOptions = getPlatformLoggerOptions(platform);

  const mixin = composeMixins([
    ...mixins,
    createContextMixin(cls),
    traceCaller && callerMixin,
  ]);

  const pinoInstance = stream
    ? pino(
        {
          ...defaultLoggerOptions,
          ...platformLoggerOptions,
          ...userPinoOpts,
          mixin,
        },
        stream,
      )
    : pino({
        ...defaultLoggerOptions,
        ...platformLoggerOptions,
        ...userPinoOpts,
        mixin,
      });

  return Object.create(pinoInstance, {
    cls: {
      value: cls,
      configurable: false,
    },
  });
}

export function expressLogger(
  namespace: Namespace,
  contextId: ClsContext['_contextId'],
  context?: ClsContext['_context'],
): RequestHandler {
  return (req, res, next): void => {
    res.set('x-context-id', contextId);

    namespace.run(() => {
      namespace.set('_contextId', contextId);
      if (context) {
        namespace.set('_context', context);
      }
      next();
    });
  };
}

export function lambdaLogger<T extends (...args: any[]) => any>(
  namespace: Namespace,
  contextId: ClsContext['_contextId'],
  context?: ClsContext['_context'],
): (next: T) => (fn: T) => ReturnType<T> {
  return (fn) => {
    return namespace.runAndReturn(() => {
      namespace.set('_contextId', contextId);
      if (context) {
        namespace.set('_context', context);
      }
      return fn();
    });
  };
}

export function createHttpLogger(
  logger: pino.Logger,
  opts: Omit<pinoHttp.Options, 'logger' | 'customLogLevel'> = {},
): pinoHttp.HttpLogger {
  return pinoHttp({
    ...opts,
    genReqId: () => ({}),
    logger,
    customSuccessMessage(res: ServerResponse) {
      return `${res.statusCode} ${res.statusMessage}`;
    },
    customErrorMessage(err: Error, res: ServerResponse) {
      return `${res.statusCode} ${res.statusMessage}`;
    },
    customLogLevel(res: ServerResponse, err: Error) {
      if (res.statusCode >= 500 || err) {
        return 'error';
      }
      if (res.statusCode >= 400 && res.statusCode < 500) {
        return 'warn';
      }
      return 'trace';
    },
    reqCustomProps(req: IncomingMessage) {
      return { url: req.url };
    },
  });
}
