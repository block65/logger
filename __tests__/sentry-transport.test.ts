import {
  afterEach,
  beforeEach,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';
import * as sentryModule from '@sentry/node';
import { LogLevelNumbers } from '../lib/types.js';

const captureException = jest.fn((exception, captureContext) => {
  return 'yes';
});

const captureMessage = jest.fn((message, captureContext) => {
  return 'yes';
});

jest.unstable_mockModule('@sentry/node', () => {
  return {
    ...sentryModule,
    captureException,
    captureMessage,
  };
});

describe('Sentry Transport', () => {
  beforeEach(() => {
    jest.resetModules();
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date('2009-02-13T23:31:30.000Z'));
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('captureException arguments', async () => {
    const { default: sentryTransport } = await import(
      '../lib/sentry-transport.js'
    );
    const { writeLogsToStream } = await import('./helpers.js');

    const transport = sentryTransport({
      dsn: 'welp',
      context: {
        user: {
          id: '1234',
        },
        tags: {
          admin: true,
        },
      },
    });

    const mockFn = jest.fn();
    transport.on('data', mockFn);

    const fakeError = new Error('Fake');

    await writeLogsToStream(transport, {
      level: LogLevelNumbers.Fatal,
      time: Date.now(),
      err: {
        message: fakeError.message,
        stack: fakeError.stack,
      },
    });

    // await new Promise(setImmediate);

    expect(captureException).toBeCalledTimes(1);

    expect(captureException.mock.calls).toMatchSnapshot();
  });
});
