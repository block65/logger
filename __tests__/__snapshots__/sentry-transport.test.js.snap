// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sentry  Logger 1`] = `
Array [
  Object {
    "err": Object {
      "message": "fake",
      "name": "Error",
      "stack": "Error: fake
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/sentry-transport.test.js:71:22)",
    },
    "hostname": "testhost",
    "level": 50,
    "msg": "fake",
    "pid": 1337,
    "time": 1234567890000,
  },
  Object {
    "err": Object {
      "message": "fake",
      "name": "Error",
      "stack": "Error: fake
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/sentry-transport.test.js:72:22)",
    },
    "hostname": "testhost",
    "level": 50,
    "msg": "fake",
    "pid": 1337,
    "time": 1234567890000,
  },
  Object {
    "err": Object {
      "message": "fake",
      "name": "Error",
      "stack": "Error: fake
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/sentry-transport.test.js:73:22)",
    },
    "hostname": "testhost",
    "level": 50,
    "msg": "fake",
    "pid": 1337,
    "time": 1234567890000,
  },
  Object {
    "err": Object {
      "message": "fake",
      "name": "Error",
      "stack": "Error: fake
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/sentry-transport.test.js:77:22)",
    },
    "hostname": "testhost",
    "level": 50,
    "msg": "fake",
    "pid": 1337,
    "time": 1234567890000,
  },
  Object {
    "err": Object {
      "message": "fake",
      "name": "Error",
      "stack": "Error: fake
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/sentry-transport.test.js:78:22)",
    },
    "hostname": "testhost",
    "level": 50,
    "msg": "fake",
    "pid": 1337,
    "time": 1234567890000,
  },
  Object {
    "err": Object {
      "message": "fake",
      "name": "Error",
      "stack": "Error: fake
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/sentry-transport.test.js:79:22)",
    },
    "hostname": "testhost",
    "level": 50,
    "msg": "fake",
    "pid": 1337,
    "time": 1234567890000,
  },
]
`;

exports[`Sentry  Transport 1`] = `
Array [
  Array [
    [Error: Oh its a fake error],
    Object {
      "level": "error",
      "tags": Object {
        "admin": true,
      },
      "user": Object {
        "id": "1234",
      },
    },
  ],
]
`;
